### PR TITLE
Add --ha-card-margin

### LINKED ISSUE #3929: Update Themes and Theme's Scope or properties
*[move[bot]](https://github.com/apps/move) commented on Nov 23, 2018, 3:46 PM UTC:*

_[marcpabst](https://github.com/marcpabst) commented on Nov 23, 2018, 8:13 AM UTC:_

Currently it's not possible to use CSS ot themes to set `border-radius` and some other properties of ha-card because encapsulation of web components works petty well in modern browsers and those properties don't use CSS custom properties as of now.

I'm therefore proposing to add a number of properties to `ha-card.js` and `ha-style.js`:

```
–card-border-radius
–card-margin
–card-border-width
–card-border-color
–card-border-style

```

This would not break anything because _if not set,_ then the properties would just resort to default values. If you agree I would be happy to make a PR.

_This issue was moved by [iantrich](https://github.com/iantrich) from [home-assistant/home-assistant-polymer#2086](https://github.com/home-assistant/home-assistant-polymer/issues/2086)._

*This issue was moved by [iantrich](https://github.com/iantrich) from [home-assistant/ui-schema#195](https://github.com/home-assistant/ui-schema/issues/195).*

### PR DESCRIPTION
**Pull request by KTibow**
## Proposed change

This adds a new variable, `--ha-card-margin`.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
frontend:
  themes:
    test_theme:
      ha-card-margin: 20px
```

## Additional information

- This PR fixes or closes issue: fixes #3929
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-card.ts [hunk 1]
```diff
@@ -32,6 +32,10 @@ export class HaCard extends LitElement {
         display: block;
         transition: all 0.3s ease-out;
         position: relative;
+        margin: var(
+         --ha-card-margin, 
+         var(--card-margin, 0px)
+        );
```
Comment (zsarnett) [inline src/components/ha-card.ts:38]: why the --card-margin? We shouldnt create two new vars?

Should just be?  right
```suggestion
        margin: var(
         --ha-card-margin, 
         0px
        );
```

Comment (KTibow) [reply src/components/ha-card.ts:38]:
```suggestion
        margin: var(--ha-card-margin, 0px);
```

**Review by zsarnett — APPROVED** [review]

**Comment by bramkragten** [comment]
Why do you want this? Theming should not set margins of the children, that is for the parents to decide. ha-card is used in so many places that this probably will cause trouble somewhere.

**Comment by KTibow** [comment]
Box shadow trouble for some themes:
image

**Comment by bramkragten** [comment]
I'm going to close this. The margin is something the parent should manage.

**Comment by KTibow** [comment]
> parent should manage

What does that mean?

**Comment by bramkragten** [comment]
The element that contains the card.

**Comment by KTibow** [comment]
You mean the custom element (the `hui-whatever-card`)? What part of the code would I change for that?

**Comment by bramkragten** [comment]
If you mean Lovelace that would be:
```
.column > * {
    display: block;
    margin: 4px 4px 8px;
}
```

**Pull request closed without merging**